traits/trait_types.py

```diff
@@ -3336,6 +3338,144 @@ def as_ctrait(self):
         """
         return self.trait_maker.as_ctrait()
 
+# -------------------------------------------------------------------------------
+#  'None' trait:
+# -------------------------------------------------------------------------------
+
+
+class NoneTrait(TraitType):
+    """ Defines a trait that only accepts the None value
+    """
+
+    info_text = "None"
+
+    default_value = None
+
+    default_value_type = DefaultValue.constant
+
+    def __init__(self, **metadata):
+        default_value = metadata.pop("default", None)
+        if default_value is not None:
+            raise ValueError("Cannot set default value {} "
+                             "for NoneTrait".format(default_value))
+        super(NoneTrait, self).__init__(**metadata)
+
+    def validate(self, obj, name, value):
+        if value is None:
+            return value
+
+        self.error(obj, name, value)
+
+# -------------------------------------------------------------------------------
+#  'Union' trait:
+# -------------------------------------------------------------------------------
+
+
+class Union(TraitType):
+    """ Defines a trait whose value can be any of of a specified list of
+    trait types or list of trait type instances or None
+    """
+    # CTrait type map for special trait types:
+    type_map = {"event": TraitKind.event, "constant": TraitKind.constant}
+
+    def __init__(self, *traits, **metadata):
+        self.list_ctrait_instances = []
+        traits = traits or [NoneTrait]
```

Comment:
```suggestion
        traits = traits if traits else [NoneTrait]
```
Fairly strong preference for this (also OK with `traits if len(traits) > 0 else [NoneTrait]`).  Short-circuiting `or` is an old idiom from before Python grew if-expressions and was a hack even back then.

Reply:
Even the `if-else` seems unnecessarily obscure here, particularly since it ends up reassigning `traits` to itself in the common case. I'd suggest something more direct, like:
```
if not traits:
    traits = [NoneTrait]
```

Reply:
There's also some type confusion here: `traits` is a tuple, while `[NoneTrait]` is a list. The types should be consistent, whether than means using `(NoneType,)`, or converting `traits` from a tuple to a list.

Reply:
Went with:
```
if not traits:
    traits = (NoneTrait,)
```